serde_json_path_core/src/node.rs

```diff
@@ -209,6 +211,391 @@ impl<'a> NodeList<'a> {
+    /// let value = json!({"foo": ["bar", "baz"]});
+    /// let path = JsonPath::parse("$.foo[0, 0, 1, 1]")?;
+    /// let mut nodes = path.query_located(&value);
+    /// assert_eq!(4, nodes.len());
+    /// nodes.dedup_in_place();
+    /// assert_eq!(2, nodes.len());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn dedup_in_place(&mut self) {
+        // This unwrap should be safe, since the paths corresponding to
+        // a query against a Value will always be ordered.
+        self.0
+            .sort_unstable_by(|a, b| a.loc.partial_cmp(&b.loc).unwrap());
+        self.0.dedup();
+    }
+
+    /// Return the first entry in the [`LocatedNodeList`], or `None` if it is empty
+    ///
+    /// # Usage
+    /// ```rust
+    /// # use serde_json::json;
+    /// # use serde_json_path::JsonPath;
+    /// # use serde_json_path::LocatedNode;
+    /// # fn main() -> Result<(), serde_json_path::ParseError> {
+    /// let value = json!({"foo": ["bar", "baz"]});
+    /// let path = JsonPath::parse("$.foo.*")?;
+    /// let nodes = path.query_located(&value);
+    /// let first = nodes.first().map(LocatedNode::node);
+    /// assert_eq!(first, Some(&json!("bar")));
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn first(&self) -> Option<&LocatedNode<'a>> {
+        self.0.first()
+    }
+
+    /// Return the last entry in the [`LocatedNodeList`], or `None` if it is empty
+    ///
+    /// # Usage
+    /// ```rust
+    /// # use serde_json::json;
+    /// # use serde_json_path::JsonPath;
+    /// # use serde_json_path::LocatedNode;
+    /// # fn main() -> Result<(), serde_json_path::ParseError> {
+    /// let value = json!({"foo": ["bar", "baz"]});
+    /// let path = JsonPath::parse("$.foo.*")?;
+    /// let nodes = path.query_located(&value);
+    /// let last = nodes.last().map(LocatedNode::node);
+    /// assert_eq!(last, Some(&json!("baz")));
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn last(&self) -> Option<&LocatedNode<'a>> {
+        self.0.last()
+    }
+
+    /// Returns the node at the given index in the [`LocatedNodeList`], or `None` if the
+    /// given index is out of bounds.
+    ///
+    /// # Usage
+    /// ```rust
+    /// # use serde_json::json;
+    /// # use serde_json_path::JsonPath;
+    /// # use serde_json_path::LocatedNode;
+    /// # fn main() -> Result<(), serde_json_path::ParseError> {
+    /// let value = json!({"foo": ["bar", "biz", "bop"]});
+    /// let path = JsonPath::parse("$.foo.*")?;
+    /// let nodes = path.query_located(&value);
+    /// assert_eq!(nodes.get(1).map(LocatedNode::node), Some(&json!("biz")));
+    /// assert!(nodes.get(4).is_none());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn get(&self, index: usize) -> Option<&LocatedNode<'a>> {
+        self.0.get(index)
+    }
+}
+
+impl<'a> From<Vec<LocatedNode<'a>>> for LocatedNodeList<'a> {
+    fn from(v: Vec<LocatedNode<'a>>) -> Self {
+        Self(v)
+    }
+}
```

Comment:
We still have this `from` impl and the TODO indicates that this can be unsafe holds.